## <a href='https://projecteuler.net/problem=26'>26. Reciprocal cycles</a>
A unit fraction contains 1 in the numerator. The decimal representation of the unit fractions with denominators 2 to 10 are given: 

$ \frac{1}{2} = 0.5 $<br>
$ \frac{1}{3} = 0.\dot3 $<br>
$ \frac{1}{4} = 0.25 $<br>
$ \frac{1}{5} = 0.2 $<br>
$ \frac{1}{6} = 0.1\dot6 $<br>
$ \frac{1}{7} = 0.\dot14285\dot7 $<br>
$ \frac{1}{8} = 0.125 $<br>
$ \frac{1}{9} = 0.\dot1 $<br>
$ \frac{1}{10} = 0.1 $<br>

Where $0.1\dot6$ means $0.166666...$, and has a 1-digit recurring cycle. It can be seen that $\frac{1}{7}$ has a 6-digit recurring cycle.

Find the value of $d < 1000$ for which $\frac{1}{d}$ contains the longest recurring cycle in its decimal fraction part.
___

> observed the above example,  <br>
1. an odd number (or even a prime) should give a longer recurring cycle (proof required) <br><br>
2. since 2/14 = 1/7 ==> i guess that for 1/d if d is multiple of a d_0 , the recurring cycle preserved (ofc proof required) <br><br>
3. after reading https://en.wikipedia.org/wiki/Repeating_decimal: 
    - The period (recurring cycle) of 1/d for integer d is always ≤ d − 1 (again proof required)
    - For those d such that 1/d have exactly d − 1 period (recurring cycle) are highly related to Cyclic number <br>
    link: https://en.wikipedia.org/wiki/Cyclic_number <br>
    or, in other word, if d is a Cyclic number, 1/d have exactly d − 1 period (recurring cycle)

> so i guess, if i can work out the largest Cyclic number such that the leading prime is under 1000, may solved the problem <br>
rmb, those found Cyclic numbers so that 1/(that prime) can have exactly d − 1 period 

> there is a formula for Cyclic number: 
$$ \frac{b^{p-1}-1}{p} $$
where b is the base of number (like base-10) and p is a prime <br><br>
this formula can be derived from:
1. Fermat quotient <br>
    https://en.wikipedia.org/wiki/Fermat_quotient <br><br>
2. Fermat's little theorem <br>
    https://en.wikipedia.org/wiki/Fermat%27s_little_theorem

> in fact: the primes that are both full reptend primes and safe primes will give the answer of this question: <br>
- full reptend primes: <br>
https://en.wikipedia.org/wiki/Full_reptend_prime <br>
- safe primes: <br>
https://en.wikipedia.org/wiki/Safe_prime <br><br>
the maximum full reptend prime under 1000 is 983, which is the answer <br>
https://oeis.org/A000353

> nah those method too hard, <br>
and i realised i need more decimal places

> well, not just too hard  
but also too many decimal places to handle  
which is a nightmare and limitation of numerical computation

> i will come back later, <br>
this is the first question that i found the answer while researching it. <br>
https://oeis.org/A051626 <br>
https://oeis.org/A051626/b051626.txt

> another attempt: (23/09) <br>
why don't we try from the back, like from 1000, 999, 998 ... <br>
once i have the number that creates exactly (n-1) decimals, then done. <br>
and such numbers must be a prime, so testing from 997, 991, 983 ...

> but still, the floating point calculation makes a problem, <br>
gotta solve them 

> the decimal approach is not feasible,  
so i removed most of the old work

> update 30jul2020  
after reading some of the post in projecteuler about this problem,  
if digging more into high-precision decimal places and floating point calculation,  
is a trap. (well it s do-able but not really practical)  
if think hard,  
there are 2(3) links(hints): 
1. fractions with prime numbers as denominator usually give more decimal places in the decimal representation recurring cycle
2. for any number that is not a multiple of the chosen denominator, the reminder when dividing that number by the chosen denominator, tells something about the decimal representation recurring cycle
3. for point 2, it s a bit like doing the Long division arithmetics, but this time count the nmumber of time when we see a repeat of reminder. 

> well, just take a look of the examples  
for $ \frac{1}{7} = 0.\dot14285\dot7 $  
we know there is 6-digit recurring cycle  
lets have a look of the reminder when dividing a number (non multiple of 7) by 7  
by long division, we choose 1/7 (this lowest number with the same number of digits, i guess would be good to go)

In [226]:
import numpy as np
import math
import time
import decimal

In [227]:
denominator = 7
dividend = 1

cycle = []
reminders = []

digit = 0    
# if 0 happens, then can't find cycle as the dividend is a multiple of denominator


while True:    # just to turn on the loop
    
    # no matter what, do the arithmetics
    digit = dividend//denominator
    reminder = dividend%denominator
    
    # if there is a repeat in remainder, stop
    if reminder in reminders:
        break
    
    # otherwise, preceed to the next long division
    else:
        cycle.append(digit)
        reminders.append(reminder)
        dividend = reminder * 10

len(cycle), cycle

(6, [0, 1, 4, 2, 8, 5])

> yup this gives a good sign

In [228]:
# function this
def recurring_digit(denominator, dividend):
    
    # storage
    cycle = []
    reminders = []
    
    digit = 0    
    # if 0 happens, then can't find cycle as the dividend is a multiple of denominator
    
    while True:    # just to turn on the loop
    
        # no matter what, do the arithmetics
        digit = dividend//denominator
        reminder = dividend%denominator
    
        # if there is a repeat in remainder, stop
        if reminder in reminders:
            break
    
        # otherwise, preceed to the next long division
        else:
            cycle.append(digit)
            reminders.append(reminder)
            dividend = reminder * 10
            
    return len(cycle)

In [229]:
limit = 1000
ans = [0,0]

t1 = time.time()
for i in range(1,1000):
    if recurring_digit(i, 100) > ans[1]:
        ans[1] = recurring_digit(i , 100)
        ans[0] = i

print(time.time()-t1, 'seconds taken')
ans

0.30422067642211914 seconds taken


[983, 982]

> yup, 983  
with 982 digit of recurring cycle

> recurring_digit(), output might be different when using a different dividend,  
but if the denominator is an odd prime and the dividend coprime denominator,  
the output should stay the same  
also, it seems like the recurring_digit() is somewhat linked to the Carmichael function  
https://en.wikipedia.org/wiki/Carmichael_function